firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -198,4 +201,84 @@ void shutdown() {
       Thread.currentThread().interrupt();
     }
   }
+
+  /**
+   * Monitors the connectivity state of the gRPC channel and resets the channel when gRPC fails to
+   * connect.
+   *
+   * <p>We currently cannot configure timeouts in connection attempts for gRPC
+   * (https://github.com/grpc/grpc-java/issues/1943), and until they support doing so, the gRPC
+   * connection can stay open for up to 2+ minutes before shutting down.
+   *
+   * <p>We start a timer when the channel enters ConnectivityState.CONNECTING. If the timer elapses,
+   * we reset the channel by shutting it down and reinitializing the channelTask. Changes to the
+   * connectivity state will clear the timer and start a new one-time listener for the next
+   * ConnectivityState change.
+   *
+   * @param channel The channel to monitor the connectivity state of.
+   */
+  private void onConnectivityStateChange(ManagedChannel channel) {
+    ConnectivityState newState = channel.getState(true);
+    Logger.debug(LOG_TAG, "Current gRPC connectivity state: " + newState);
+    // Check that the new state is online, then cancel timer.
+    if (newState == ConnectivityState.CONNECTING) {
+      Logger.debug(LOG_TAG, "Setting the connectivityAttemptTimer");
+      hardAssert(
+          connectivityAttemptTimer == null,
+          "connectivityAttemptTimer should be null when setting a new timer.");
```

Comment:
How confident are you in this assert? 😁 

The docs for [notifyWhenStateChanged](https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannel.html#notifyWhenStateChanged-io.grpc.ConnectivityState-java.lang.Runnable-) say:

> There is an inherent race between the notification to callback and any call to getState(). There is a similar race between getState() and a call to notifyWhenStateChanged(). The state can change during those races, so there is not a way to see every state transition. "Transitions" to the same state are possible, because intermediate states may not have been observed. The API is only reliable in tracking the current state.

So I think that means we could end up seeing CONNECTING twice in a row, and this assert would fail and the SDK would crash.

So in this case we probably need to be defensive (just clear the existing timer if it exists) rather than strict (throwing an error if the timer already exists).

Reply:
Not very after reading those docs! Thanks for catching this!